feat: [TR-458] note model

JoacoGDR opened the pull request:
## JIRA Card

https://widergy.atlassian.net/browse/TR-458

## Summary

Se implementó el modelo de Note, para que los usuarios puedan crear notas acerca de los libros que tienen asociados. Los campos que se prentenden agregar son los siguientes:

- _title: String (Requerido)_
                
- _content: String (Requerido)_
                
- _note_type: String (Requerido) (review o critique)_
                
- _user: references  (Required)_
                
Para eso se creo un nuevo archivo de migracion para crear la tabla en la base de datos y el modelo de Note. Ademas, se modifico el modelo de Users para definir la relacion de one-to-many que tiene con Note. 

No se generaron controladores para Note, por lo tanto la documentacion de la API no se altero.


## Screenshots
En el screenshot se puede ver como se crea una nota y la misma se puede acceder desde el User.
![image](https://github.com/user-attachments/assets/52c2a2ad-21e9-44b0-af0c-d6dcba74ce2d)

A continuacion screens de output de los tests y linter

![image](https://github.com/user-attachments/assets/180bb64c-0d43-4c31-9895-f1eda3652946)

## Test Cases
Desde la consola de Rails, crear una nota de este modo:
`Note.create!(title: 'Saludos!', content: 'Esta es una reseña', note_type: :review, user_id: 1)`

Luego, verificar que el User pueda acceder a ella de esta manera:

```
user = User.find(1)
user.notes # Debe devolver las notas que creo el usuario
```

JoacoGDR committed:
feat: note model and migration

Created Note table through a migration and its model

JoacoGDR committed:
feat: user has_many notes

Added the one to many relation to user model. Also fixed the order of fields in the note model.

JoaquinHetrea commented on db/migrate/20240726172356_create_notes.rb:
```diff
@@ -0,0 +1,11 @@
+    create_table :notes do |t|
+      t.string :title, null:false
+      t.text :content, null:false
+      t.string :note_type, null:false
+      t.references :user, foreign_key: true
+      t.timestamps
+    end
+  end
```
La migración es correcta, pero nos está faltando un detalle :eyes:. Revisemos que las validaciones del modelo se correspondan con las de la nueva tabla

JoaquinHetrea commented on app/models/note.rb:
```diff
@@ -0,0 +1,19 @@
+class Note < ApplicationRecord
+  belongs_to :user
+  has_one :utility, through: :user
+  validates :title, presence: true
+  validates :content, presence: true
+  validates :note_type, presence: true
```
Excelentes estas validaciones!! Usemos las herramientas de rails para dejarlo en una sola linea como charlamos

JoaquinHetrea commented on app/models/note.rb:
```diff
@@ -0,0 +1,19 @@
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+class Note < ApplicationRecord
```
No te olvides de sumar un archivo de factories para el nuevo modelo. Aprovecho para dejarte la documentación de Faker para que chusmees. Es muy completa y nos ayuda muchisimo en los tests: https://github.com/faker-ruby/faker

JoaquinHetrea reviewed (commented):
Excelente Joaco!! Muy buen trabajo. Revisemos unos puntitos antes de finalizar!!

Nuevamente muy buena la documentación del PR!! :clap: :clap: No te olvides de sumar unas capturitas corriendo los tests y rubocop :robot: :police_officer:

JoacoGDR committed:
fix: reverted changes to Gemfile.lock

JoacoGDR committed:
fix: consistent validations and factory tests

Defined Factory Bot model and added test to note_spec. Updated seed file to populate a few notes. Fixed data type of content in migration and relation consistencies with model.

JoaquinHetrea commented on db/migrate/20240729144500_create_notes.rb:
```diff
@@ -0,0 +1,12 @@
+      t.string :title, null: false
+      t.string :content, null: false
+      t.string :note_type, null: false
+      t.belongs_to :user, foreign_key: true
```
Esta columna debería ser opcional? Revisemos lo que dice la card, lo que valida el modelo y las restricciones en la bdd que genera esta migración

JoaquinHetrea commented on app/models/note.rb:
```diff
@@ -0,0 +1,17 @@
+class Note < ApplicationRecord
+  belongs_to :user
+  has_one :utility, through: :user
+  validates :user_id, :title, :content, :note_type, presence: true
```
Muy bien!! Aún nos falta una validación. Se puede actualmente crear una nota indicando un `note_type` arbitrario? Se debería poder eso? Voy dejando un enlace que te va a resultar útil: https://blog.saeloun.com/2022/01/05/how-to-use-enums-in-rails/

JoaquinHetrea commented on spec/models/note_spec.rb:
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+
+RSpec.describe Note, type: :model do
+  subject(:note) do
+    create(:note)
+  end
+
```
Buen adelanto :clap: :clap: más adelante vamos a tener una card especificamente para trabajar los tests del modelo de notas. Por el momento estos tests son más que suficientes :rocket:

JoaquinHetrea reviewed (commented):
Buen avance Joaco! Dejo unos puntitos más para ir revisando, ya casi estamos!! :muscle:

JoacoGDR committed:
fix: note_type enum

Changed note_type to enum. Validated user_id presence too.

JoacoGDR committed:
fix: outdated note_spec fixed

JoacoGDR committed:
fix: note_type presence

JoaquinHetrea reviewed (approved):
Buen laburo Joaco!! aquí tiene su approve: :white_check_mark: :rocket: 

No te olvides de **actualizar el estado de las cards** y **actualizar las capturas de los tests** con la info más reciente!!

:rocket:
